src/firmware_new/boards/frankie_v1/io/ublox_odinw262_communicator.c

```diff
@@ -0,0 +1,209 @@
+    // https://github.com/UBC-Thunderbots/Software/issues/1875
+    io_ublox_odinw262_communicator_sendATCommand("AT+UMLA=2,00AAAAAAAA00\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT&W\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+CPWROFF\r");
+    io_ublox_odinw262_communicator_waitForBoot();
+
+    io_ublox_odinw262_communicator_sendATCommand("AT+UBRGC=0,0,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UBRGC=0,1,3,1\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UBRGC=0,100,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UBRGC=0,107,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UBRGCA=0,3\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UETHC=0,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UETHC=1,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UETHC=2,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UETHC=3,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UETHC=5,3\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UETHC=4,1\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UETHCA=3\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UWSC=0,0,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UWSC=0,2,\"" WIFI_SSID "\"\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UWSC=0,5,2\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UWSC=0,8,\"" WIFI_PASS "\"\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UWSC=0,300,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UWSC=0,301,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UWSCA=0,3\r");
+
+    for (;;)
+    {
+        osDelay(1000);
+    }
+}
+
+void io_ublox_odinw262_communicator_handleIdleLine()
+{
+    g_dma_counter_on_uart_idle_line =
+        RX_BUFFER_LENGTH_BYTES - __HAL_DMA_GET_COUNTER(g_ublox_uart_handle->hdmarx);
+    g_ublox_response_status =
+        io_ublox_odinw262_communicator_getUbloxResponseStatusFromCircularBuffer(
+            g_dma_counter_on_uart_idle_line, RX_BUFFER_LENGTH_BYTES,
+            g_dma_uart_receive_buffer);
+    osSemaphoreRelease(g_dma_receive_semaphore);
+}
+
+void io_ublox_odinw262_reset()
+{
+    assert(g_initialized);
+    io_gpio_pin_setActive(g_ublox_reset_pin);
+    io_gpio_pin_setInactive(g_ublox_reset_pin);
+}
+
+void io_ublox_odinw262_communicator_waitForBoot(void)
+{
+    do
+    {
+        TLOG_INFO("Waiting for u-blox to boot up");
+        io_ublox_odinw262_communicator_sendATCommand("AT\r");
+    } while (g_ublox_response_status != UBLOX_RESPONSE_OK);
+
+    TLOG_INFO("u-blox detected");
+}
+
+char* io_ublox_odinw262_communicator_sendATCommand(const char* command)
+{
+    assert(g_initialized);
+    TLOG_INFO("Sending AT Command to u-blox: %s", command);
+
+    HAL_UART_Transmit(g_ublox_uart_handle, (uint8_t*)command, (uint16_t)strlen(command),
+                      HAL_MAX_DELAY);
+
+wait_for_ublox_to_respond:
```

Comment:
can this be a function?

Reply:
Bumping

Reply:
Ah this got folded away mb, this is a bit hard because we want the function to jump up to the semaphore every-time we get an incomplete response so we grab the appropriate data when it does come through. 

But if the sempahore pend times out we want to return. I'm not sure how I can logically break this into a function and not have the `goto`